src/k8s-engine/lima.ts

```diff
@@ -1362,6 +1367,34 @@ ${ commands.join('\n') }
             }
           });
 
+        // We can't install buildkitd earlier because if we were running an older version of rancher-desktop,
+        // we have to remove the kim buildkitd k8s artifacts. And we can't remove them until k8s is running.
+        // Note that if the user's workflow is:
+        // A. Only containerd
+        // settings version 3: containerd (which installs buildkitd)
+        // upgrade to settings version 4, still on containerd:
+        //   - remove the old kim/buildkitd artifacts
+        //   - set config.kubernetes.checkForExistingKimBuilder to false (forever)
+
+        // B. Mix of containerd and moby
+        // settings version 3: containerd (which installs buildkitd)
+        // settings version 3: switch to moby (which will uninstall buildkitd)
+        // upgrade to settings version 4, still on moby: do nothing here
+        // settings version 4, switch to containerd
+        //   - config.kubernetes.checkForExistingKimBuilder should be true, but there are no kim/buildkitd artifacts
+        //   - do nothing, and set config.kubernetes.checkForExistingKimBuilder to false (forever)
+
+        if (config.checkForExistingKimBuilder) {
+          this.client ??= new K8s.Client();
+          await getImageProcessor(this.#currentContainerEngine, this).removeKimBuilder(this.client.k8sClient);
+          // No need to remove kim builder components ever again.
+          config.checkForExistingKimBuilder = false;
+          this.emit('kim-builder-uninstalled');
```

Comment:
I believe these two lines to the same thing (`config` is a reference to the settings object), but it's fine to have both.

Reply:
Actually, the upgrade to `config` is for the current code, while the message persists the changes to the config. It turns out that currently nothing else checks `config.checkForExistingKimBuilder`, but I'd rather leave the line that sets it to `false` right now for emphasis.